### PR TITLE
GODRIVER-3240 Code hardening.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-3240

## Summary
- Use functions from the "encoding/binary" package to replace bit-shifting logic as a follow-up of #1646.
- Update size check for integer conversions.
- Add allocation size check.
- Convert atlas test to a gotest.

### CODE REVIEW

File: bson/bsonrw/extjson_wrappers.go [hunk 1]
```diff
@@ -95,12 +96,14 @@ func (ejv *extJSONValue) parseBinary() (b []byte, subType byte, err error) {
...
-			i, err := strconv.ParseInt(val.v.(string), 16, 64)
+			i, err := strconv.ParseUint(val.v.(string), 16, 64)
 			if err != nil {
 				return nil, 0, fmt.Errorf("invalid $binary subType string: %s", val.v.(string))
 			}
 
-			subType = byte(i)
+			b := []byte{0, 0, 0, 0, 0, 0, 0, 0}
+			binary.LittleEndian.PutUint64(b, i)
+			subType = b[0]
```
Comment (matthewdale) [inline bson/bsonrw/extjson_wrappers.go:106]: We can use `ParseUint` to limit the range of the parsed subtype value by using `bitSize=8`. I also recommend adding the parse error to returned error.

E.g.
```go
i, err := strconv.ParseUint(val.v.(string), 16, 8)
if err != nil {
	return nil, 0, fmt.Errorf("invalid $binary subType string %q: %w", val.v.(string), err)
}
subType = byte(i)
```

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: bson/bsoncodec/uint_codec.go [hunk 2]
```diff
@@ -164,11 +164,15 @@ func (uic *UIntCodec) decodeType(dc DecodeContext, vr bsonrw.ValueReader, t refl
 
 		return reflect.ValueOf(uint64(i64)), nil
 	case reflect.Uint:
-		if i64 < 0 || int64(uint(i64)) != i64 { // Can we fit this inside of an uint
+		if i64 < 0 {
+			return emptyValue, fmt.Errorf("%d overflows uint", i64)
```
Comment (matthewdale) [inline bson/bsoncodec/uint_codec.go:168]: Use a more accurate error message.

```suggestion
			return emptyValue, fmt.Errorf("%d underflows uint", i64)
```

**Review by matthewdale — CHANGES_REQUESTED** [review]

File: mongo/options/clientoptions.go [hunk 3]
```diff
@@ -1177,7 +1178,19 @@ func addClientCertFromSeparateFiles(cfg *tls.Config, keyFile, certFile, keyPassw
...
+	keySize := len(keyData)
+	if keySize > 64*1024*1024 {
+		return "", errors.New("X.509 key must be less than 64 MiB")
+	}
+	certSize := len(certData)
+	if certSize > 64*1024*1024 {
+		return "", errors.New("X.509 certificate must be less than 64 MiB")
+	}
+	dataSize := keySize + certSize + 1
+	if dataSize > math.MaxInt {
```
Comment (qingyang-hu) [inline mongo/options/clientoptions.go:1190]: An additional fail-safe check.

**Comment by blink1073** [comment]
The `TestDefaultValueDecoders` test failures are relevant.

File: bson/bsoncodec/uint_codec.go (hunk 2 above)
Comment (qingyang-hu) [reply bson/bsoncodec/uint_codec.go]: It seems better to keep the error message consistent with other unsigned types.

Comment (matthewdale) [reply bson/bsoncodec/uint_codec.go:168]: Sounds reasonable. We can fix them all at the same time. Consider this comment resolved.

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

**Pull request merged**